In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline


In [2]:
df = pd.read_csv('cardekho_imputated.csv')


In [3]:
df.head()

Unnamed: 0       car_name    brand     model  vehicle_age  km_driven  \
0           0    Maruti Alto   Maruti      Alto            9     120000   
1           1  Hyundai Grand  Hyundai     Grand            5      20000   
2           2    Hyundai i20  Hyundai       i20           11      60000   
3           3    Maruti Alto   Maruti      Alto            9      37000   
4           4  Ford Ecosport     Ford  Ecosport            6      30000   

  seller_type fuel_type transmission_type  mileage  engine  max_power  seats  \
0  Individual    Petrol            Manual    19.70     796      46.30      5   
1  Individual    Petrol            Manual    18.90    1197      82.00      5   
2  Individual    Petrol            Manual    17.00    1197      80.00      5   
3  Individual    Petrol            Manual    20.92     998      67.10      5   
4      Dealer    Diesel            Manual    22.77    1498      98.59      5   

   selling_price  
0         120000  
1         550000  
2         215000  
3         226000  
4         570000

In [4]:
df.drop(columns = 'Unnamed: 0',inplace=True)

In [5]:
df.head()


car_name    brand     model  vehicle_age  km_driven seller_type  \
0    Maruti Alto   Maruti      Alto            9     120000  Individual   
1  Hyundai Grand  Hyundai     Grand            5      20000  Individual   
2    Hyundai i20  Hyundai       i20           11      60000  Individual   
3    Maruti Alto   Maruti      Alto            9      37000  Individual   
4  Ford Ecosport     Ford  Ecosport            6      30000      Dealer   

  fuel_type transmission_type  mileage  engine  max_power  seats  \
0    Petrol            Manual    19.70     796      46.30      5   
1    Petrol            Manual    18.90    1197      82.00      5   
2    Petrol            Manual    17.00    1197      80.00      5   
3    Petrol            Manual    20.92     998      67.10      5   
4    Diesel            Manual    22.77    1498      98.59      5   

   selling_price  
0         120000  
1         550000  
2         215000  
3         226000  
4         570000

## Data Cleaning

### handling missing values

handling missing values

check duplicates

check data types

understand the dataset

In [6]:
df.isnull().sum()

car_name             0
brand                0
model                0
vehicle_age          0
km_driven            0
seller_type          0
fuel_type            0
transmission_type    0
mileage              0
engine               0
max_power            0
seats                0
selling_price        0
dtype: int64

In [7]:
df.drop(columns = ['car_name','brand'],axis=1,inplace=True)

In [8]:
df.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      Alto            9     120000  Individual    Petrol            Manual   
1     Grand            5      20000  Individual    Petrol            Manual   
2       i20           11      60000  Individual    Petrol            Manual   
3      Alto            9      37000  Individual    Petrol            Manual   
4  Ecosport            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  selling_price  
0    19.70     796      46.30      5         120000  
1    18.90    1197      82.00      5         550000  
2    17.00    1197      80.00      5         215000  
3    20.92     998      67.10      5         226000  
4    22.77    1498      98.59      5         570000

In [9]:
len(df.axes[1])

11

In [10]:
# checking unique values

df['model'].unique()

array(['Alto', 'Grand', 'i20', 'Ecosport', 'Wagon R', 'i10', 'Venue',
       'Swift', 'Verna', 'Duster', 'Cooper', 'Ciaz', 'C-Class', 'Innova',
       'Baleno', 'Swift Dzire', 'Vento', 'Creta', 'City', 'Bolero',
       'Fortuner', 'KWID', 'Amaze', 'Santro', 'XUV500', 'KUV100', 'Ignis',
       'RediGO', 'Scorpio', 'Marazzo', 'Aspire', 'Figo', 'Vitara',
       'Tiago', 'Polo', 'Seltos', 'Celerio', 'GO', '5', 'CR-V',
       'Endeavour', 'KUV', 'Jazz', '3', 'A4', 'Tigor', 'Ertiga', 'Safari',
       'Thar', 'Hexa', 'Rover', 'Eeco', 'A6', 'E-Class', 'Q7', 'Z4', '6',
       'XF', 'X5', 'Hector', 'Civic', 'D-Max', 'Cayenne', 'X1', 'Rapid',
       'Freestyle', 'Superb', 'Nexon', 'XUV300', 'Dzire VXI', 'S90',
       'WR-V', 'XL6', 'Triber', 'ES', 'Wrangler', 'Camry', 'Elantra',
       'Yaris', 'GL-Class', '7', 'S-Presso', 'Dzire LXI', 'Aura', 'XC',
       'Ghibli', 'Continental', 'CR', 'Kicks', 'S-Class', 'Tucson',
       'Harrier', 'X3', 'Octavia', 'Compass', 'CLS', 'redi-GO', 'Glanza',
       

In [11]:
num_features = [feature for feature in df.columns if df[feature].dtype != 'O']
cat_features = [feature for feature in df.columns if df[feature].dtype == 'O']
print("number of numeric features: ",len(num_features))
print("number of categorical features: ",len(cat_features))
discrete_features = [feature for feature in num_features if len(df[feature].unique()) <= 25]
print("number of discrete features: ",len(discrete_features))
cont_features = [feature for feature in num_features if feature not in discrete_features]
print("number of continuos features: ",len(cont_features))

number of numeric features:  7
number of categorical features:  4
number of discrete features:  2
number of continuos features:  5


In [12]:
# Dependent and Independent feature split
X = df.drop(['selling_price'],axis=1)
y = df['selling_price']

In [13]:
X.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      Alto            9     120000  Individual    Petrol            Manual   
1     Grand            5      20000  Individual    Petrol            Manual   
2       i20           11      60000  Individual    Petrol            Manual   
3      Alto            9      37000  Individual    Petrol            Manual   
4  Ecosport            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  
0    19.70     796      46.30      5  
1    18.90    1197      82.00      5  
2    17.00    1197      80.00      5  
3    20.92     998      67.10      5  
4    22.77    1498      98.59      5

### feature encoding and scaling

In [14]:
len(df['model'].unique())

120

we will not do one hot encoding here coz we want to find relation of model with otput variable (price), and one hot encoding will create 119 extra columns (coz len model = 120)

In [15]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X['model'] = le.fit_transform(X['model'])

In [16]:
X.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      7            9     120000  Individual    Petrol            Manual   
1     54            5      20000  Individual    Petrol            Manual   
2    118           11      60000  Individual    Petrol            Manual   
3      7            9      37000  Individual    Petrol            Manual   
4     38            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  
0    19.70     796      46.30      5  
1    18.90    1197      82.00      5  
2    17.00    1197      80.00      5  
3    20.92     998      67.10      5  
4    22.77    1498      98.59      5

for the categories below we can do one hot encoding coz no of categories are less

In [17]:
len(df['seller_type'].unique())

3

In [18]:
len(df['fuel_type'].unique())

5

In [19]:
len(df['transmission_type'].unique())

2

pandas.core.frame.DataFrame

In [20]:
num_features = X.select_dtypes(exclude='object').columns

In [21]:
onehot_features = ['seller_type','fuel_type','transmission_type']

In [22]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
 
numeric_transformer = StandardScaler()
oh_transformer = OneHotEncoder(drop='first')

pre_processor = ColumnTransformer(
    [
        ("OneHotEncoder",oh_transformer,onehot_features),
        ("StandardScaler",numeric_transformer,num_features)
    ],remainder='passthrough'
)

# remainder = passthrough means all other columns will stay as it is, if we dont use passthrough, it will delete all rest features


In [23]:
# to see rows and cols
df.shape

(15411, 11)

In [24]:
X = pre_processor.fit_transform(X)

In [25]:
pd.DataFrame(X)

0    1    2    3    4    5    6         7         8          9   \
0      1.0  0.0  0.0  0.0  0.0  1.0  1.0 -1.519714  0.983562   1.247335   
1      1.0  0.0  0.0  0.0  0.0  1.0  1.0 -0.225693 -0.343933  -0.690016   
2      1.0  0.0  0.0  0.0  0.0  1.0  1.0  1.536377  1.647309   0.084924   
3      1.0  0.0  0.0  0.0  0.0  1.0  1.0 -1.519714  0.983562  -0.360667   
4      0.0  0.0  1.0  0.0  0.0  0.0  1.0 -0.666211 -0.012060  -0.496281   
...    ...  ...  ...  ...  ...  ...  ...       ...       ...        ...   
15406  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.508844  0.983562  -0.869744   
15407  0.0  0.0  0.0  0.0  0.0  1.0  1.0 -0.556082 -1.339555  -0.728763   
15408  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.407551 -0.012060   0.220539   
15409  0.0  0.0  1.0  0.0  0.0  0.0  1.0  1.426247 -0.343933  72.541850   
15410  0.0  0.0  0.0  0.0  0.0  1.0  0.0 -1.024131 -1.339555  -0.825631   

             10        11        12        13  
0     -0.000276 -1.324259 -1.263352 -0.403022  
1     -0.192071 -0.554718 -0.432571 -0.403022  
2     -0.647583 -0.554718 -0.479113 -0.403022  
3      0.292211 -0.936610 -0.779312 -0.403022  
4      0.735736  0.022918 -0.046502 -0.403022  
...         ...       ...       ...       ...  
15406  0.026096 -0.767733 -0.757204 -0.403022  
15407 -0.527711 -0.216964 -0.220803  2.073444  
15408  0.344954  0.022918  0.068225 -0.403022  
15409 -0.887326  1.329794  0.917158  2.073444  
15410 -0.407839  0.020999  0.395884 -0.403022  

[15411 rows x 14 columns]

In [26]:
# train test split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=10)

In [27]:
X_train.shape,X_test.shape

((12328, 14), (3083, 14))

### Model training and model selection

In [28]:
from xgboost import XGBRegressor

In [29]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.linear_model import LinearRegression,Ridge,Lasso
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score,mean_absolute_error,mean_squared_error


In [30]:
# function to evaluate model
def evaluate_model(true,predicted):
    mae = mean_absolute_error(true,predicted)
    mse = mean_squared_error(true,predicted)
    rmse = np.sqrt(mean_squared_error(true,predicted))
    r2_square = r2_score(true,predicted)
    return mae,rmse,r2_square


In [34]:
models = {
    "Linear Regression": LinearRegression(),
    "Lasso": Lasso(),
    "Ridge": Ridge(),
    "K-Neighbors Regressor": KNeighborsRegressor(),
    "DecisionTree": DecisionTreeRegressor(),
    "RandomForestRegressor": RandomForestRegressor(),
    "AdaBoostRegressor": AdaBoostRegressor(),
    "XG Boost Regresssor": XGBRegressor()
}
for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train,y_train) # train the model

    # make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluate train and test dataset

    model_train_mae, model_train_rmse, model_train_r2 = evaluate_model(y_train,y_train_pred)
    model_test_mae, model_test_rmse,model_test_r2 = evaluate_model(y_test,y_test_pred)

    print(list(models.keys())[i])
    print("Model performance for training set")
    print("Root mean sq error", round(model_train_rmse,2))
    print("mean absolute error",round(model_train_mae,2))
    print("R2 score",round(model_train_r2,2))
    print("----------------------------------------------------")
    print("\n")
    print("Model performance for test dataset")
   
    print("Root mean sq error", round(model_test_rmse,2))
    print("mean absolute error",round(model_test_mae,2))
    print("R2 score",round(model_test_r2,2))
    print('='*35)
    print('\n')





Linear Regression
Model performance for training set
Root mean sq error 478801.85
mean absolute error 260031.95
R2 score 0.68
----------------------------------------------------


Model performance for test dataset
Root mean sq error 753414.83
mean absolute error 280915.98
R2 score 0.51


Lasso
Model performance for training set
Root mean sq error 478801.85
mean absolute error 260030.05
R2 score 0.68
----------------------------------------------------


Model performance for test dataset
Root mean sq error 753414.1
mean absolute error 280914.87
R2 score 0.51


Ridge
Model performance for training set
Root mean sq error 478802.54
mean absolute error 260001.49
R2 score 0.68
----------------------------------------------------


Model performance for test dataset
Root mean sq error 753427.9
mean absolute error 280894.9
R2 score 0.51


K-Neighbors Regressor
Model performance for training set
Root mean sq error 232145.6
mean absolute error 88356.15
R2 score 0.92
--------------------------

In [35]:
# parameters for hyperpara tuning
#knn_params = {"n_neighbors": [2,3,10,20,40,50]}
rff_params = {"max_depth": [5,8,15,None,10],
              "max_features": [5,7,"auto",8],
              "min_samples_split": [2,8,15,20],
              "n_estimators": [100,200,500,1000]}
#adaboost_params = {"n_estimators": [50,60,70,80],
   #                "loss": ['linear','square','exponential']
    #               }
xg_param = { "learning_rate": [0.1,0.01],
            "max_depth": [5,8,12,20,30],
            "n_estimators": [100,200,300],
            "colsample_bytree": [0.5,0.8,1,0.3,0.4]

}




In [36]:
randomcv_models = [
                   ('RFF',RandomForestRegressor(),rff_params),
                   ('XGG',XGBRegressor(),xg_param)
]


In [37]:
from sklearn.model_selection import RandomizedSearchCV

In [38]:
model_param = {}
for name,model,params in randomcv_models:
    random = RandomizedSearchCV(estimator = model,param_distributions=params,n_iter=100,cv=3,verbose=2,n_jobs=-1)
    random.fit(X_train,y_train)
    model_param[name] = random.best_params_
    

Fitting 3 folds for each of 100 candidates, totalling 300 fits
[CV] END max_depth=None, max_features=5, min_samples_split=20, n_estimators=100; total time=   2.3s
[CV] END max_depth=None, max_features=5, min_samples_split=20, n_estimators=100; total time=   2.4s
[CV] END max_depth=None, max_features=5, min_samples_split=20, n_estimators=100; total time=   2.4s
[CV] END max_depth=8, max_features=5, min_samples_split=15, n_estimators=500; total time=   8.2s
[CV] END max_depth=8, max_features=5, min_samples_split=15, n_estimators=500; total time=   8.3s
[CV] END max_depth=8, max_features=5, min_samples_split=15, n_estimators=500; total time=   8.4s
[CV] END max_depth=5, max_features=7, min_samples_split=8, n_estimators=200; total time=   3.0s
[CV] END max_depth=15, max_features=5, min_samples_split=15, n_estimators=200; total time=   4.8s
[CV] END max_depth=15, max_features=5, min_samples_split=15, n_estimators=200; total time=   4.9s
[CV] END max_depth=15, max_features=5, min_samples_spl

/Users/user/Desktop/anaconda3/lib/python3.12/site-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
63 fits failed out of a total of 300.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
33 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/user/Desktop/anaconda3/lib/python3.12/site-packages/sklearn/model_selection/_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/user/Desktop/anaconda3/lib/python3.12/site-packages/sklearn/base.py", line 1467, in wrapper
    estimator._validate_params()
  File "/Users/user/Desktop/anaconda3/lib/python3.12/site-packages/sklearn/base.py", line 666, in _validate_params
    validate_parameter

Fitting 3 folds for each of 100 candidates, totalling 300 fits
[CV] END colsample_bytree=1, learning_rate=0.1, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END colsample_bytree=1, learning_rate=0.1, max_depth=5, n_estimators=100; total time=   0.4s
[CV] END colsample_bytree=1, learning_rate=0.1, max_depth=5, n_estimators=100; total time=   0.5s
[CV] END colsample_bytree=1, learning_rate=0.01, max_depth=12, n_estimators=300; total time=  12.3s
[CV] END colsample_bytree=0.8, learning_rate=0.01, max_depth=20, n_estimators=200; total time=  14.6s
[CV] END colsample_bytree=0.8, learning_rate=0.01, max_depth=20, n_estimators=200; total time=  14.2s
[CV] END colsample_bytree=0.8, learning_rate=0.01, max_depth=20, n_estimators=200; total time=  14.2s
[CV] END colsample_bytree=0.8, learning_rate=0.1, max_depth=12, n_estimators=200; total time=   4.5s
[CV] END colsample_bytree=0.8, learning_rate=0.1, max_depth=12, n_estimators=200; total time=   4.7s
[CV] END colsample_bytree=0.3, lear

In [39]:
model_param

{'RFF': {'n_estimators': 500,
  'min_samples_split': 2,
  'max_features': 7,
  'max_depth': 15},
 'XGG': {'n_estimators': 300,
  'max_depth': 5,
  'learning_rate': 0.1,
  'colsample_bytree': 1}}

In [40]:
# Retraining of models wth best parameters
models = {
"Random Forest Regressor": RandomForestRegressor(n_estimators= 200,
  min_samples_split= 2,
  max_features= 7,
  max_depth= 15),
#"KNeighborsRegressor": KNeighborsRegressor(n_neighbors=2),
#"adaboostAlgo": AdaBoostRegressor(n_estimators= 60, loss= 'linear')
"XG_BOOST": XGBRegressor(n_estimators=300,max_depth=5,learning_rate=0.1,colsample_bytree=1)
}
for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train,y_train) # train the model

    # make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluate train and test dataset

    model_train_mae, model_train_rmse, model_train_r2 = evaluate_model(y_train,y_train_pred)
    model_test_mae, model_test_rmse,model_test_r2 = evaluate_model(y_test,y_test_pred)

    print(list(models.keys())[i])
    print("Model performance for training set")
    print("Root mean sq error", round(model_train_rmse,2))
    print("mean absolute error",round(model_train_mae,2))
    print("R2 score",round(model_train_r2,2))
    print("----------------------------------------------------")
    print("\n")
    print("Model performance for test dataset")
    print("Model performance for training set")
    print("Root mean sq error", round(model_test_rmse,2))
    print("mean absolute error",round(model_test_mae,2))
    print("R2 score",round(model_test_r2,2))
    print('='*35)
    print('\n')


Random Forest Regressor
Model performance for training set
Root mean sq error 100019.7
mean absolute error 52756.79
R2 score 0.99
----------------------------------------------------


Model performance for test dataset
Model performance for training set
Root mean sq error 527179.42
mean absolute error 105676.29
R2 score 0.76


XG_BOOST
Model performance for training set
Root mean sq error 105125.27
mean absolute error 71040.66
R2 score 0.98
----------------------------------------------------


Model performance for test dataset
Model performance for training set
Root mean sq error 524953.68
mean absolute error 104262.31
R2 score 0.76


